react/src/layout/report.js

```diff
@@ -0,0 +1,297 @@
+                {VisualComponent && <VisualComponent key={simulationPollingVersionRef.current} config={reportLayout} simulationData={simulationData}/>}
+                {inProgress && <ProgressBar animated now={100}/>}
+            </>
+        )
+    }
+}
+
+function getFrameId({
+    frameIndex,
+    reportName,
+    simulationId,
+    appName,
+    computeJobHash,
+    computeJobSerial,
+    hookedDependendies
+}) {
+    let frameIdElements = [
+        frameIndex,
+        reportName,
+        simulationId,
+        appName,
+        computeJobHash,
+        computeJobSerial,
+        ...((hookedDependendies || []).map(v => v.value))
+    ]
+
+    return frameIdElements.join('*');
+}
+
+export class ManualRunReportLayout extends View {
+    getFormDependencies = (config) => {
+        let { reportLayout } = config;
+        let layoutElement = this.layoutsWrapper.getLayoutForConfig(reportLayout);
+        return layoutElement.getFormDependencies();
+    }
+
+    component = (props) => {
+        let { config } = props;
+        let { reportName, reportGroupName, reportLayout, frameIdFields, shown: shownConfig } = config;
+        
+        let reportEventManager = useContext(ContextReportEventManager);
+        let schema = useContext(ContextSchema);
+        let modelsWrapper = useContext(ContextModelsWrapper);
+        let simulationInfoPromise = useContext(ContextSimulationInfoPromise);
+        let appName = useContext(ContextAppName);
+
+        let [reportData, updateReportData] = useState(undefined);
+
+        let reportEventsVersionRef = useRef(uuidv4())
+
+        let shown = true;
+
+        if(shownConfig) {
+            shown = useEvaluatedInterpString(modelsWrapper, shownConfig);
+        }
+
+        let frameIdDependencies = frameIdFields.map(f => new Dependency(f));
+
+        let frameIdDependencyGroup = new HookedDependencyGroup({ 
+            dependencies: frameIdDependencies,
+            modelsWrapper,
+            schemaModels: schema.models
+        })
+
+        let hookedFrameIdDependencies = frameIdDependencies.map(frameIdDependencyGroup.getHookedDependency);
+
+        useEffect(() => {
+            let reportEventsVersion = uuidv4();
+            reportEventsVersionRef.current = reportEventsVersion;
+            
+            reportEventManager.onReportData(reportGroupName, (simulationData) => {
+                if(reportEventsVersionRef.current !== reportEventsVersion) {
+                    return; // guard against concurrency with older versions
+                }
+
+                updateReportData(undefined);
+
+                let { state } = simulationData;
+                if(state == "completed") {
+                    simulationInfoPromise.then(({simulationId}) => {
+                        let { frameCount, computeJobHash, computeJobSerial } = simulationData;
+                        let frameId = getFrameId({
+                            frameIndex: frameCount - 1, // pick last frame for now
```

Comment:
"on the side" comments" are hard to manage. Rather, name a function "lastFrame(frameCount)"